Reimplement Kaiser window

alvgaona opened the pull request:
Relates to #85366 

- For reference follow #87082.

alvgaona committed:
Add kaiser to signal.windows

alvgaona commented:
@mruberry @lezcano I'm getting this lint error.

```python
>>> Lint for torch/__init__.py:

  Error (MYPY) [no-redef]
    Name "kaiser_window" already defined (possibly by an import)

        874  |
        875  |# Importing alias functions due to backwards compatibility.
        876  |# Link to a detailed conversation: https://github.com/pytorch/pytorch/pull/87082
    >>> 877  |from torch._windows import kaiser_window
        878  |
        879  |# Quantized, sparse, AO, etc. should be last to get imported, as nothing
        880  |# is expected to depend on them.
```

I feel like we need to somehow stop generating the Python bindings for the C++ implementation for this warning to get solved. Additionally, I also feel like we should remove this block as it has old documentation and doesn't really add anything. We're doing the alias to a different implementation.

https://github.com/pytorch/pytorch/blob/c9b618447d7c948003f26c3b49c28cdc193bd3f0/torch/_torch_docs.py#L12723-L12765

One last question. I got lost a little bit in the previous conversation about what to do with the C++ implementations, the PrimTorch, and the `_refs` directory (still learning about this kind of modules).

lezcano commented:
You can stop PyTorch from generating bindings by adding a `manual_cpp_binding: True` to its corresponding entry in `native_functions.yaml`. 
About the docs, agree. You should replace those docs with docs of other functions that are alias, as Mike suggested in the previous PR.
About the refs, it may be fine not to do anything on that end really (unless @mruberry thinks otherwise). Already having this implemented in Python is a massive step forward, and adding these to the `_refs` folder (which, as you may have already seen, is a folder with Python implementations of many PyTorch functions) should be fairly straightforward and can be done at a later time.

mruberry commented on torch/signal/windows/windows.py:
```diff
@@ -296,3 +297,83 @@ def gaussian(
+    ),
+)
+def kaiser(
+        M: int,
```
So this module is moving some of the positional SciPy arguments to be kwarg-only, but it's also providing defaults for them? In particular, `std` in `gaussian` above has a new default, and so does `beta` in `kaiser` here.

lezcano replied:
IMO, if these defaults are reasonable (and a-priori they are), I'm happy with that, but I'll let @alvgaona assess.

mruberry replied:
Not saying we shouldn't do it, just verifying my understanding

alvgaona replied:
Yes, we're making it to kwarg-only as well as giving it a default value. The SciPy implementations aren't consistent in this matter; if you look at `scipy.signal.windows.gaussian`, `std` is a required arg which is completely different from `scipy.signal.windows.exponential` for the parameter `tau`, so we're improving in that sense. 

In terms of kwarg-only, I believe that this kind of function is used with hardcoded values most of the time so to make the call a little bit more readable, enforcing the parameter name in the call sounds very reasonable.

For instance, 

```python
torch.signal.windows.gaussian(10, 2)
```

isn't as clear as

```python
torch.signal.windows.gaussian(10, std=2)
```

You could argue that the first argument is a required arg, but a convention is that the first parameter is always the size of the window, which is easy to remember.

mruberry commented on torch/testing/_internal/opinfo/definitions/signal.py:
```diff
@@ -95,6 +95,23 @@ def reference_inputs_gaussian_window(op_info, device, dtype, requires_grad, **kw
         yield SampleInput(size, sym=True, **kw)
 
 
+def reference_inputs_kaiser_window(op_info, device, dtype, requires_grad, **kwargs):
```
Nice reference inputs

mruberry commented on torch/testing/_internal/opinfo/definitions/signal.py:
```diff
@@ -156,6 +173,18 @@ def error_inputs_gaussian_window(op_info, device, **kwargs):
     )
 
 
+def error_inputs_kaiser_window(op_info, device, **kwargs):
```
Nice error inputs

mruberry commented on torch/testing/_internal/opinfo/definitions/signal.py:
```diff
@@ -310,4 +339,13 @@ def make_signal_windows_opinfo(
         reference_inputs_func=partial(reference_inputs_gaussian_window, std=1.92),
         error_inputs_func=error_inputs_gaussian_window,
     ),
+    make_signal_windows_opinfo(
```
Cool

mruberry commented on torch/_windows.py:
```diff
@@ -0,0 +1,30 @@
+__all__ = ["kaiser_window"]
+
+
+def kaiser_window(
```
My suggestion is not to worry about `torch.kaiser_window` in this PR.

A follow-up PR will let us focus on the `torch.signal.windows.kaiser` logic, and `torch.kaiser_window` isn't really an alias for `torch.signal.windows.kaiser` because of the `window_length` and `periodic` arguments having different names

alvgaona replied:
Ok, I'll undo this change. Just for the record, the alias we'd like to have the same signature as the old one, correct?

mruberry replied:
Yes. More precisely it would be a "near alias" that forwards its arguments to this implementation

mruberry commented on torch/signal/windows/windows.py:
```diff
@@ -296,3 +297,83 @@ def gaussian(
+
+@_add_docstr(
+    r"""
+Computes the Kaiser window
```
"window."

mruberry commented on torch/signal/windows/windows.py:
```diff
@@ -296,3 +297,83 @@ def gaussian(
+.. math::
+    out_i = I_0 \left( \beta \sqrt{1 - \left( {\frac{i - N/2}{N/2}} \right) ^2 } \right) / I_0( \beta )
+
+Where :math:`I_0` is the zeroth order modified Bessel function of the first kind (see :func:`torch.i0`) and
```
link to the canonical `torch.special.i0` instead of the alias (saving the reader a click)

mruberry commented on torch/signal/windows/windows.py:
```diff
@@ -296,3 +297,83 @@ def gaussian(
+    out_i = I_0 \left( \beta \sqrt{1 - \left( {\frac{i - N/2}{N/2}} \right) ^2 } \right) / I_0( \beta )
+
+Where :math:`I_0` is the zeroth order modified Bessel function of the first kind (see :func:`torch.i0`) and
+``N = M - 1`` if :attr:`sym` is True and ``M`` if :attr:`sym` is False,
```
ug, our doc's rendering of mathematical expressions like this is horrible. I would just remove the annotations from the math and write "N = M - 1 if sym is True and M if sym is False"

lezcano replied:
My kingdom for a half-decent sphinx theme.

alvgaona replied:
So @mruberry you don't like `:math:`I_0``, right? Personally, I think it looks kinda cool.

alvgaona replied:
@lezcano @mruberry can we put it like this? I think it's 1000% more readable than the current one:

![image](https://user-images.githubusercontent.com/13088001/197431310-afcd9e6c-870e-4c5a-83da-8acb8725f8a8.png)

For comparison, here's the current one 🤮 

![image](https://user-images.githubusercontent.com/13088001/197431468-83111ffa-2ad9-46ac-abce-2207369a15aa.png)

lezcano replied:
I like the special link. It's also true that our rendering of double ticks in Sphinx is less than good. Now, simple ticks should be a good compromise. The thing with LaTeX is that it takes some time to render, so that's why we tend to use it just for large equations.
Also, I think that rather than just saying yhat when yhe window is symmetric something happens, I like pointing the user to the relevant arg.

alvgaona replied:
I think I prefer using double ticks. I removed the `:math:` annotation and applied back your last suggestion.

mruberry reviewed (commented):
This looks great; I would just leave `torch.kaiser_window` as-is for the moment and focus on the development of `torch.signal.windows`. (That thought goes for the development of additional windows, too). I made some small inline comments about `torch.signal.windows.kaiser` for your review. 

We can review aligning the existing `torch` windowing operations with `torch.signal.windows` later.

alvgaona reviewed (commented):
@mruberry addressed a couple of your comments with additional comments 😛

alvgaona commented:
@lezcano do we want to reimplement the C++ impl with linspace? I think you've mentioned something like that.

lezcano commented:
I did mention we may want to update the C++ implementations, but I think it's alright to leave them as they are for now, so that we don't generate unnecessary extra work.

mruberry commented on torch/__init__.py:
```diff
@@ -872,6 +872,10 @@ def _assert(condition, message):
 
+# Importing alias functions due to backwards compatibility.
+# Link to a detailed conversation: https://github.com/pytorch/pytorch/pull/87082
+from torch._windows import kaiser_window
```
This line might be causing the test failures?

alvgaona commented:
> You can stop PyTorch from generating bindings by adding a `manual_cpp_binding: True` to its corresponding entry in `native_functions.yaml`. About the docs, agree. You should replace those docs with docs of other functions that are alias, as Mike suggested in the previous PR. About the refs, it may be fine not to do anything on that end really (unless @mruberry thinks otherwise). Already having this implemented in Python is a massive step forward, and adding these to the `_refs` folder (which, as you may have already seen, is a folder with Python implementations of many PyTorch functions) should be fairly straightforward and can be done at a later time.

@lezcano I tried using `manual_cpp_binding: True` but it fails. I guess we can have a look at the follow-up PR where we try to do the aliasing work.

lezcano commented:
follow up PR LGTM!

alvgaona committed:
Fix docstring

alvgaona committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into kaiser-window

alvgaona commented:
@lezcano @mruberry I think we can merge it and go back to #87082, right?

alvgaona committed:
Fix docstring

ngimel commented on torch/signal/windows/windows.py:
```diff
@@ -296,3 +297,84 @@ def gaussian(
+    if M == 1:
+        return torch.ones((1,), dtype=dtype, layout=layout, device=device, requires_grad=requires_grad)
+
+    if beta <= 0:
```
check for beta should go before returning the wrong value for M=1

mruberry replied:
What's interesting is that SciPy does not require a positive beta:

```
scipy.signal.windows.kaiser(1, beta=-5)
: array([1.])

scipy.signal.windows.kaiser(10, beta=-5.)
: array([0.03671089, 0.20127873, 0.47552746, 0.7753221 , 0.97273069,
       0.97273069, 0.7753221 , 0.47552746, 0.20127873, 0.03671089])
```

The wikipedia entry stipulates that beta should be non-negative, but not strictly positive: https://en.wikipedia.org/wiki/Kaiser_window

So, follow-up questions:

-  why are we requiring a positive beta?
- should the documentation discuss this requirement?

ngimel replied:
Sure, let's make it non-negative. Definition includes I0(beta) and I0 is not defined for negative args (and there's no alternative definition for 1), so looks like scipy is wrong.

alvgaona replied:
Agreed with @ngimel. I'll turn that precheck into a non-negative one.

alvgaona replied:
@lezcano @mruberry @ngimel I guess for the gaussian and exponential windows we should move the `std` and `tau` check up a few lines too.

lezcano replied:
yup. Feel free to add it to this PR if oyu want.

alvgaona replied:
I'll do it in #87082

mruberry replied:
> Agreed with @ngimel. I'll turn that precheck into a non-negative one, and I'll move it up as well.

But why not just allow negative betas, too?

alvgaona replied:
@mruberry 

![image](https://user-images.githubusercontent.com/13088001/198110555-65dbeaa0-b8f1-4516-8c98-a1056e138bdb.png)

mruberry replied:
I understand, just interesting that SciPy allows them

lezcano commented on torch/signal/windows/windows.py:
```diff
@@ -296,3 +297,84 @@ def gaussian(
+                       device=device,
+                       requires_grad=requires_grad)
+
+    return torch.i0(beta * torch.sqrt(1 - torch.pow(k, 2))) / torch.i0(torch.tensor(beta))
```
Nit. You can golf this down by one kernel by putting the beta inside the square root and doing sqrt(b*b - (b*k)**2) and then fusing the b *k into the creation of k.

alvgaona replied:
Sure! Good catch

lezcano reviewed (approved):
Leaving aside a small point I left and Natalia's valid point, this LGTM. Let's see what Mike has to say though.

As always, thank you for doing all this ground work!

alvgaona committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into kaiser-window

alvgaona committed:
Update code regarding beta

- Make the precheck non-negative.
- Call fewer kernels by adding beta in the construction of k

alvgaona commented:
@lezcano I think I forgot not to branch out from `master`, that's why tests are failing, right?

alvgaona committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into kaiser-window

mruberry commented on torch/signal/windows/windows.py:
```diff
@@ -296,3 +297,84 @@ def gaussian(
+    {M}
+
+Keyword args:
+    beta (float, optional): shape parameter for the window. Default: 12.0
```
Let's note this should be non-negative, too

mruberry reviewed (approved):
See one suggestion about mentioning non-negative requirement for beta in the documentation, otherwise LGTM! Thanks @alvgaona

alvgaona committed:
Update docstring for beta

alvgaona committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into kaiser-window

alvgaona commented:
@mruberry @lezcano ready to get merged. Is it possible to do a squash merge?

lezcano commented:
@pytorchbot merge
